src/rustup-dist/src/download.rs

```diff
@@ -1,95 +1,186 @@
-
-            let target_url = try!(utils::parse_url(url));
-            let target_file = try!(self.temp_cfg.new_file());
-
-            {
-                let target_filename: &Path = &target_file;
-                try!(utils::download_file(&target_url,
-                                          &target_file,
-                                          None,
-                                          &|n| (self.notify_handler)(n.into())));
-
-                let key_file = try!(self.temp_cfg.new_file());
-                let key_filename: &Path = &key_file;
-                try!(utils::write_file("key", &key_file, key));
-
-                let gpg = try!(utils::find_cmd(&["gpg2", "gpg"])
-                               .ok_or("could not find 'gpg' on PATH"));
-
-                try!(utils::cmd_status("gpg",
-                                       Command::new(gpg)
-                                           .arg("--no-permission-warning")
-                                           .arg("--dearmor")
-                                           .arg(key_filename)));
-
-                try!(utils::cmd_status("gpg",
-                                       Command::new(gpg)
-                                           .arg("--no-permission-warning")
-                                           .arg("--keyring")
-                                           .arg(&key_filename.with_extension("gpg"))
-                                           .arg("--verify")
-                                           .arg(target_filename)));
+
+    /// Downloads a file, validating its hash, and resuming interrupted downloads
+    /// Partial downloads are stored in `self.download_dir`, keyed by hash. If the
+    /// target file already exists, then the hash is checked and it is returned
+    /// immediately without re-downloading. 
+    pub fn download(&self, url: &Url, hash: &str) -> Result<File> {
+
+        try!(utils::ensure_dir_exists("Download Directory", &self.download_dir, &|n| (self.notify_handler)(n.into())));
+        let target_file = self.download_dir.join(Path::new(hash));
+
+        if target_file.exists() {
+            let cached_result = try!(file_hash(&target_file));
+            if hash == cached_result {
+                (self.notify_handler)(Notification::FileAlreadyDownloaded);
+                (self.notify_handler)(Notification::ChecksumValid(&url.to_string()));
+                return Ok(File { path: target_file, });
+            } else {
+                (self.notify_handler)(Notification::CachedFileChecksumFailed);
+                try!(fs::remove_file(&target_file).chain_err(|| "cleaning up previous download"));
             }
+        }
+
 
-            Ok(target_file)
+        let partial_file_path =
+            target_file.with_file_name(
+                target_file.file_name().map(|s| {
+                    s.to_str().unwrap_or("_")})
+                    .unwrap_or("_")
```

Comment:
Hm, I know you didn't add this, but falling back to underscore seems like a bad idea, and the filename is supposed to be the hash anyway, so this should be an `expect` instead.